### PR TITLE
217 218 229 230 Big refactoring and cleanup

### LINKED ISSUE #217: SAMP Controller Refactor
Incl. Updating the status checker.

### LINKED ISSUE #218: Document New SAMP Controller Behavior

### LINKED ISSUE #229: Remove Thread.sleep from the BuilderMainViewTest
By putting something in that lets us wait for completion of the spawned Events' threads.

### LINKED ISSUE #230: Jacoco Build Issue
Fix an issue with Jacoco and incorrect/broken coverage analysis

### PR DESCRIPTION
**Pull request by olaurino**
This PR builds on previous PRs/branches and introduces refactorings and fixes to make sure the SAMP testing issues are fixed (at least most of the time).

Several changes are embedded in the updates to sherpa-samp.

Replaces #228
Resolve #217
Resolve #218
Resolve #229 
Resolve #230

### CODE REVIEW

**Comment by jbudynk** [comment]
I'm still reviewing the code, but a a few general things I noticed while testing out the branch:

1.  On OS X, once in a while `mvn clean verify` or `mvn clean package` will fail at RedshiftInterpolateIT. 2/8 times of running either `mvn clean verify` or `mvn clean package` this happened.

2.  On Linux and OS X: If sherpa gets disconnected from the SAMP hub, the Sherpa-samp connection icon stays "connected." It only disconnects when Iris disconnects from the SAMP hub. I tried using SedStacker when I knew sherpa was disconnected, and the operations failed as expected.

**Comment by jbudynk** [comment]
Command line arguments "`iris --test`" and "`iris --vizier`" no longer bring up a desktop icon for me (on OS X). I've run `mvn package` and `mvn install`; neither build brings up a TestBuilder or VizieR client on the desktop. 

I think the issue is in "Iris.java::initialize()" at [lines 122-125](https://github.com/ChandraCXC/iris/pull/231/files#diff-30fe77a8dbe4f4ef4d901626e96fe337R122). It doesn't look like CommandLine components get loaded with the ComponentLoader or with Configuration().apply().

File: iris/src/main/java/cfa/vo/iris/Iris.java [hunk 1]
```diff
@@ -15,35 +15,71 @@
...
 import javax.swing.JDialog;
-import javax.swing.UIManager;
 
 /**
  * The main class of the application.
  */
-public class Iris extends AbstractIrisApplication {
+public class Iris extends Application implements IrisApplication {
+
+    private final Logger logger = Logger.getLogger(Iris.class.getName());
```
Comment (eholum) [inline iris/src/main/java/cfa/vo/iris/Iris.java]: static

File: samp-factory/src/main/java/cfa/vo/interop/SAMPServiceBuilder.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    /**
+     * Set whether the controller should start and internal HTTP server
+     * @param serverRoot the root path of the resource server
+     * @return
+     */
+    public SAMPServiceBuilder withResourceServer(String serverRoot) {
+        this.withResourceServer = true;
+        if (serverRoot != null) {
```
Comment (eholum) [inline samp-factory/src/main/java/cfa/vo/interop/SAMPServiceBuilder.java]: This is set above, so the serverRoot will never actually change?

File: samp-factory/src/main/java/cfa/vo/interop/SampService.java [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class SampService {
+    private final String PING_MTYPE = "samp.app.ping";
```
Comment (eholum) [inline samp-factory/src/main/java/cfa/vo/interop/SampService.java]: statics

File: samp-factory/src/main/java/cfa/vo/interop/SampService.java [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        started = false;
+    }
+
+    public void addSampConnectionListener(SAMPConnectionListener listener) {
+        sampListeners.add(listener);
+        listener.run(sampUp);
+    }
+
+    private void monitorSampOnce() {
+        logger.log(Level.INFO, "Monitor State: ");
```
Comment (eholum) [inline samp-factory/src/main/java/cfa/vo/interop/SampService.java]: I might put just one line at the INFO level, and these other ones at debug, this is a lot of info to dump out every second.

File: samp-factory/src/main/java/cfa/vo/sherpa/SherpaClient.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+public class SherpaClient {
+
+    private SampService sampService;
+    private Map<String, AbstractModel> modelMap = new HashMap<>();
+    private Integer stringCounter = 0;
+    private Logger logger = Logger.getLogger(SherpaClient.class.getName());
```
Comment (eholum) [inline samp-factory/src/main/java/cfa/vo/sherpa/SherpaClient.java]: static

**Comment by olaurino** [comment]
As we agreed, I'll merge this PR. Some issues are still open, but we have more tasks for them:
  * If sherpa disconnects, the desktop icon still shows it connected (#233).
  * Some components (we know about SedStacker for sure) freeze the GUI during a sherpa call. With the new retry code in place, this means that the GUI is frozen for 10 seconds before Iris gives up. We need to a) make sure those components do not freeze the GUI, and b) the user is aware that something is wrong but we are retrying. (#234)
  * Erratically, some tests will fail. So far we have seen that happening with the Redshift Integration Test, and we may want to investigate some more (#235).

There was also an issue with some command line arguments being ignored, but that's fixed by #241, which was already incorporated in this PR.

**Pull request merged**